File: gan/ALinear.py
import torch
import torch.nn as nn
import random
import pandas as pd
import time
class LinearRegressionModel(nn.Module):
    def __init__(self, input_dim, output_dim):
        super(LinearRegressionModel, self).__init__()
        self.linear = nn.Linear(input_dim, output_dim)

    def forward(self, x):
        return self.linear(x)
    
    def configure_optimizers(self):
        return torch.optim.SGD(self.parameters(), 0.03)
    
    def loss(self, y_hat, y):
        fn = nn.MSELoss()
        return fn(y_hat, y)
    
class Dataset:
    def __init__(self, train_size, test_szie):
        w = torch.tensor([2, -3.4])
        b = 4.2
        noise = 0.01 

        self.train_size = train_size
        self.test_size = test_szie

        n = train_size + test_szie
        self.x = torch.randn(n, len(w))
        noise = torch.randn(n, 1) * noise
        self.y = torch.matmul(self.x, w.reshape((-1, 1))) + b + noise
        #self.y = y.reshape([len(y)])
    
    def get_train_data(self):
        x_train = self.x[range(0, self.train_size)]
        y_train = self.y[range(0, self.train_size)]
        return (x_train, y_train)
    
    def get_test_data(self):
        x_test = self.x[range(self.train_size, -1)]
        y_test = self.y[range(self.train_size, -1)]
        return (x_test, y_test)
    
    def save(self, file):
        data = torch.cat((self.x, self.y), 1)
        pf = pd.DataFrame(data)
        pf.to_csv(file, index = False, header=["x0", "x1", "y"])


class Learner:
    def __init__(self, epochs, batch_size, lr = 0.0005):
        self.epochs = epochs
        self.batch_size = batch_size
        self.lr = lr

    def fit(self, model, data):
        #data = Dataset(1024, 128)
        x_train, y_train = data.get_train_data()
        x_test, y_test = data.get_test_data()

        sgd = model.configure_optimizers()
        n_batch = len(x_train) // self.batch_size
        idxs = list(range(0, len(x_train))) 
        random.shuffle(idxs)

        total_time = 0
        for epoch in range(self.epochs):
            total_loss = 0.0

            start_time = time.perf_counter()
            for i in range(n_batch):
                x = x_train[range(i * self.batch_size, i * self.batch_size + self.batch_size)]
                y = y_train[range(i * self.batch_size, i * self.batch_size + self.batch_size)]
                #x = x.to_device()
                #y = y.to_device()

                y_hat = model.forward(x)
                loss = model.loss(y_hat, y)
                sgd.zero_grad()
                loss.backward()
                sgd.step()

                total_loss += loss
        
            end_time = time.perf_counter()
            diff = end_time - start_time
            total_time += diff
            avg_loss = total_loss / n_batch
            print("Epoch consumes ", diff)
            print("Epoch {} -- avg_loss: {}\n".format(epoch, avg_loss))
        
        print("Cosume time :", total_time )
        y_hat = model(torch.tensor([1.0, 1.0]))
        print(y_hat)

if __name__ == "__main__":
    d = Dataset(1024, 128)
    d.save("./../dataset/regression.csv")
    x_train, y_train = d.get_train_data()
    x_test, y_test = d.get_test_data()
    
    m = LinearRegressionModel(2, 1)

    leaner = Learner(10, 128) 
    leaner.fit(m, d)

File: rust/candle-guide/src/simple_linear/a_linear.rs
use rand::prelude::SliceRandom;
use rand::rng;
use crate::simple_linear::dataset::Dataset;
use super::*;
use candle_core::{DType, Result, Tensor};
use candle_nn::{loss::mse, optim, Linear, Module, Optimizer, VarBuilder, VarMap};
use std::time::{Duration, Instant};

// a very import lesson here is that a Linear perception should have two parts: w and b
fn linear_z(in_dim: usize, out_dim: usize, vs: VarBuilder)->Result<Linear>{
    let ws = vs.get_with_hints((out_dim, in_dim), "weight",candle_nn::init::ZERO)?;
    let bs = vs.get_with_hints(out_dim, "bias", candle_nn::init::ZERO)?;
    Ok(Linear::new(ws, Some(bs)))
}
struct SLP{
    ln: Linear,
}

impl SLP {
    pub fn new(vs: VarBuilder) -> Result<Self> {
        //let ln = candle_nn::linear(2, 1, vs.pp("ln"))?;
        let ln = linear_z(2, 1, vs)?;
        Ok(Self {ln})
    }

    pub fn forward(&self, x: &Tensor) -> Result<Tensor> {
        let x = self.ln.forward(x)?;
        Ok(x) 
    }

    pub fn validate(&self)->Result<()>{
        let x_valid_vec: Vec<f64> = vec![
            0.9, 0.5
        ];

        let x_valid = Tensor::from_vec(x_valid_vec.clone(), (1, 2), &device)?.to_dtype(DType::F64)?.to_device(&device)?;
        let y_valid = self.forward(&x_valid)?;
        
        //println!("x_valid: {:?}", x_valid.to_vec0::<f32>()?);
        println!("y_valid: {:?}", y_valid.reshape(())?.to_scalar::<f64>()?);
        Ok(())
    }
}
pub struct Learner{
    epochs: usize,
    batch_size: usize,
    lr: f64
}

impl Learner{
    pub fn new(epochs: usize, batch_size: usize) -> Learner{
        let lr = 0.0005;
        Learner {epochs, batch_size, lr}
    }

    pub fn train(&self, m: Dataset) -> Result<()> {
        let mut varmap = VarMap::new();
        let vs = VarBuilder::from_varmap(&varmap, DType::F64, &device);
        let model = SLP::new(vs.clone())?;
        let mut sgd = optim::SGD::new(varmap.all_vars(), self.lr)?;

        let x_train = m.x_train.to_device(&device)?.to_dtype(DType::F64)?.to_device(&device)?;
        let y_train = m.y_train.to_device(&device)?.to_dtype(DType::F64)?.to_device(&device)?;
        let x_test = m.x_test.to_device(&device)?.to_dtype(DType::F64)?;
        let y_test = m.y_test.to_device(&device)?.to_dtype(DType::F64)?;
        let n = x_train.dim(0)?;
        let nbatches = n / self.batch_size;
        let mut idxs = (0..nbatches).collect::<Vec<usize>>();
        
        let mut total_time = 0u128;
        for epoch in 0..self.epochs{
            let mut sum_loss = 0.0f64;
            idxs.shuffle(&mut rng());
            let start = Instant::now();
            for idx in idxs.iter(){
                let _x = x_train.narrow(0, idx*self.batch_size, self.batch_size)?;
                let _y = y_train.narrow(0, idx*self.batch_size, self.batch_size)?;
                let y_hat = model.forward(&_x)?;
                let loss = mse(&y_hat, &_y)?;               
                sgd.backward_step(&loss)?;
                sum_loss += loss.to_vec0::<f64>()?;               
            }
            let diff = start.elapsed().as_micros();
            println!("Epoch consumes: {}", diff);
            total_time += diff;
            let avg_loss = sum_loss / (nbatches as f64);
            let y_hat = model.forward(&x_test)?;
            //let accuracy = model.loss(&y_hat, &y_test)?;
            let accuracy = mse(&y_hat, &y_test)?.to_vec0::<f64>()? / (y_test.dim(0)? as f64); 
            println!("Epoch {:?} -- avg_loss:{:?}, avg_accuracy:{:?}\n", epoch, avg_loss, accuracy);     
        }
        println!("Consume {:?} microseconds totoally.", total_time);
        let _ = model.validate();
        Ok(())
    }
}